README.md
# PHPCalendarApi

[![CI workflow](https://github.com/bjoern-hempel/php-calendar-api/actions/workflows/ci-workflow.yml/badge.svg?branch=main)](https://github.com/bjoern-hempel/php-calendar-api/actions/workflows/ci-workflow.yml)
[![Release](https://img.shields.io/github/v/release/bjoern-hempel/php-calendar-api)](https://github.com/bjoern-hempel/php-calendar-api/releases)
[![PHP](https://img.shields.io/badge/PHP-^8.0-777bb3.svg?logo=php&logoColor=white&labelColor=555555&style=flat)](https://www.php.net/supported-versions.php)
[![PHPStan](https://img.shields.io/badge/PHPStan-Level%208-brightgreen.svg?style=flat)](https://phpstan.org/user-guide/rule-levels)
[![PHPCS](https://img.shields.io/badge/PHPCS-PSR12-brightgreen.svg?style=flat)](https://www.php-fig.org/psr/psr-12/)
[![LICENSE](https://img.shields.io/github/license/bjoern-hempel/php-calendar-api)](https://github.com/bjoern-hempel/php-calendar-api/blob/master/LICENSE.md)

> An API for building ready-made calendars. Includes a user area and token management. The framework is secured via tests and static code analyzers.

# 0. Main Technologies

## 0.1 Applications and Extensions

* **Symfony 6.1.x**: https://symfony.com/releases/6.1
    * **Released**: May 2022
    * **Support until**: January 2023
    * https://symfony.com/releases/6.1
    * https://endoflife.date/symfony
* **API Platform 3.0.x**: https://api-platform.com/docs
    * **Released**: 15 September 2022
    * https://github.com/api-platform/api-platform/releases
    * https://endoflife.date/api-platform

## 0.2 Scripting language and database management systems

* **PHP 8.1.x**: https://www.php.net/releases/8.1/en.php
    * **Released**: 23 November 2021
    * **Active Support Until**: 25 November 2023
    * **Security Support Until**: 25 November 2024
    * https://www.php.net/supported-versions.php
    * https://endoflife.date/php
* **MariaDB 10.7.3**: https://mariadb.com/kb/en/mariadb-1073-release-notes/
    * **Released**: 8 February 2022
    * **Support until**: 14 February 2023
    * https://endoflife.date/mariadb

## 0.3 Helper

* **Composer 2.4.x**: https://getcomposer.org/download
* **Symfony Client**: https://symfony.com/download

## 0.4 Testing and Analysis

* **PHPUnit 9.5.x**: https://phpunit.de
* **PHPStan: 1.9.x**: https://phpstan.org
* **PHP Magic Number Detector (PHPMND): 3.0.x**: https://github.com/povils/phpmnd
* **Rector PHP: 0.14.x**: https://github.com/rectorphp/rector

## 0.5 Manuals

* **Integration**: https://symfony.com/doc/current/the-fast-track/de/26-api.html

## 0.6 Techniques

* **JSONLD**: https://en.wikipedia.org/wiki/JSON-LD
* **JSON**: https://en.wikipedia.org/wiki/JSON
* **REST**: https://en.wikipedia.org/wiki/Representational_state_transfer

# 1. Installation

```bash
git clone git@github.com:bjoern-hempel/calendarBuilder.git && cd calendarBuilder
```

## 1.1 Local development

The development of the unit classes, the unit tests and the general framework can be done locally.

### 1.1.1 Check PHP version

At least version 8.1 is required.

```bash
php -v
```

```bash
PHP 8.1.12 (cli) (built: Oct 28 2022 18:35:51) (NTS)
Copyright (c) The PHP Group
Zend Engine v4.1.12, Copyright (c) Zend Technologies
    with Zend OPcache v8.1.12, Copyright (c), by Zend Technologies
    with Xdebug v3.1.5, Copyright (c) 2002-2022, by Derick Rethans
```

### 1.1.2 Composer install

```bash
composer install
```

### 1.1.3 Show version

```bash
bin/console version:show
```

```bash
  Version:         0.33.0
  License:         Copyright (c) 2022 Björn Hempel
  Authors:         Björn Hempel <bjoern@hempel.li>
  PHP Version:     8.1.12
  Symfony Version: 6.1.7
```

### 1.1.4 Reinitialize the test database

Required for functional tests:

```bash
bin/console db:reinitialize --env=test
```

or

```bash
composer db:initialize:test
```

Check db size:

```bash
du -h var/cache/test/test.db
```

```text
232K    var/cache/test/test.db
```

### 1.1.5 Run first test

```bash
composer test
```

## 1.2 Execution with docker

If there is no local PHP version 8.1, or you want to develop the API with real data and a database, the Docker LAMP
the integrated Docker LAMP stack can be used. This procedure is the recommended way.

> **Attention**: This procedure needs an external Traefik network. If this has not yet been done, it must be created beforehand:
>
> ```bash
> docker network create traefik-public
> ```

### 1.2.1 Start LAMP stack

```bash
docker compose pull && docker compose up -d
```

### 1.2.2 Composer install

Use the user `user` to keep the permissions:

```bash
docker compose exec -u user php composer install
```

### 1.2.3 Show version

#### Command Line

```bash
docker compose exec -u user php bin/console version:show
```

```bash
  Version:         0.33.0
  License:         Copyright (c) 2022 Björn Hempel
  Authors:         Björn Hempel <bjoern@hempel.li>
  PHP Version:     8.1.12
  Symfony Version: 6.1.7
```

### 1.2.4 Reinitialize the test and development database

Required for functional tests (test database):

```bash
docker compose exec -u user php bin/console db:reinitialize --env=test
```

Check db size:

```bash
docker compose exec -u user php du -h var/cache/test/test.db
```

```text
232K    var/cache/test/test.db
```

### 1.2.5 Run first test

```bash
docker compose exec -u user php composer test
```

## 1.3 Install dependencies

```bash
docker compose exec php /etc/init.d/supervisor start
```

```bash
docker compose exec php composer install
```

```bash
docker compose exec php yarn install
```

```bash
docker compose exec php yarn encore production
```

```bash
docker compose exec php composer migrate-prod
```

# 2. Requirements

* PHP 8.1

## 2.1 PHP Modules

```bash
php -m
```

```bash
[PHP Modules]
bcmath
calendar
Core
ctype
curl
date
dom
exif
FFI
fileinfo
filter
ftp
gettext
hash
iconv
imagick
intl
json
libxml
mbstring
mysqli
mysqlnd
openssl
pcntl
pcre
PDO
pdo_mysql
pdo_sqlite
Phar
posix
readline
Reflection
session
shmop
SimpleXML
sockets
sodium
SPL
sqlite3
standard
sysvmsg
sysvsem
sysvshm
tokenizer
xdebug
xml
xmlreader
xmlwriter
xsl
Zend OPcache
zlib

[Zend Modules]
Xdebug
Zend OPcache
```

# 3. API Documentation

* [API - v1](docs/api/README.md)

# 4. Run tests

Test DB preparations if needed:

```bash
bin/console db:reinitialize --env=test
```

Check sqlite db:

```bash
ls -la var/app.db
```

## 4.1 General tests

Runs all available tests (phpmnd, phpunit, phpstan, phpcs, rector):

```bash
composer test
```

or

```bash
composer test:all
```

Runs basic tests (phpunit, phpstan):

```bash
composer test:basic
```

Runs most tests (phpmnd, phpunit, phpstan, phpcs):

```bash
composer test:most
```

## 4.2 PHP Coding Standards Fixer

* **Web**: https://cs.symfony.com/
* **Github**: https://github.com/FriendsOfPHP/PHP-CS-Fixer
* **Wikipedia**: https://en.wikipedia.org/wiki/PHP_Standard_Recommendation

We use the PSR-12 standard rule in this project.

### 4.2.1 Dry-Run

Runs PHP Coding Standards Fixer for all folders (dry-run) and check for inconsistencies. No sources are changed in
the process:

```bash
composer phpcs:check:all
```

Runs PHP Coding Standards Fixer for the src folder (dry-run) and check for inconsistencies. No sources are changed in
the process:

```bash
composer phpcs:check:src
```

Runs PHP Coding Standards Fixer for the tests folder (dry-run) and check for inconsistencies. No sources are changed in
the process:

```bash
composer phpcs:check:tests
```

### 4.2.2 Fix source code

Runs PHP Coding Standards Fixer for all folders (fix). The source code is tried to be adapted according to the
defined rules:

```bash
composer phpcs:fix:all
```

Runs PHP Coding Standards Fixer for the src folder (fix). The source code is tried to be adapted according to the
defined rules:

```bash
composer phpcs:fix:src
```

Runs PHP Coding Standards Fixer for the tests folder (fix). The source code is tried to be adapted according to the
defined rules:

```bash
composer phpcs:fix:tests
```

## 4.3 PHP Mess Detector

* **Web**: https://phpmd.org/
* **Github**: https://github.com/phpmd/phpmd
* **Other**: https://phpqa.io/projects/phpmd.html

To get more information on this topic, run the following command:

```bash
composer | grep "phpmd:"
```

## 4.4 PHP Magic Number Detector

* **Github**: https://github.com/povils/phpmnd
* **Wikipedia**: https://en.wikipedia.org/wiki/Magic_number_(programming)

Runs PHP Magic Number Detector and detects magic numbers for all folders:

```bash
composer phpmnd:all
```

Runs PHP Magic Number Detector and detects magic numbers for the src folder:

```bash
composer phpmnd:src
```

Runs PHP Magic Number Detector and detects magic numbers for the tests folder:

```bash
composer phpmnd:tests
```

## 4.5 PHPStan

* **Web**: https://phpstan.org/
* **Github**: https://github.com/phpstan/phpstan
* **Wikipedia**: https://en.wikipedia.org/wiki/Static_program_analysis

```bash
composer phpstan:run
```

or

```bash
composer phpstan
```

## 4.6 PHPUnit

* **Web**: https://phpunit.de/
* **Github**: https://github.com/sebastianbergmann/phpunit

Runs all available PHPUnit tests (api, functional, unit):

```bash
composer phpunit
```

or

```bash
composer phpunit:all
```

Runs only unit PHPUnit tests:

```bash
composer phpunit:api
```

Runs only functional PHPUnit tests:

```bash
composer phpunit:functional
```

Runs only unit PHPUnit tests:

```bash
composer phpunit:unit
```

## 4.7 PHP Rector

* **Web**: https://getrector.org/
* **Github**: https://github.com/rectorphp/rector

Runs PHP Rector and does a dry-run:

```bash
composer rector:check
```

Runs PHP Rector and fix source code:

```bash
composer rector:fix
```

# 5. Docker Compose

## 5.1 Basic Commands

### 5.1.1 Start LAMP stack

```bash
docker compose up -d
```

### 5.1.2 Stop LAMP stack

```bash
docker compose down
```

### 5.1.3 Rebuild images

```bash
docker compose build
```

### 5.1.4 Load dumps into database

All SQL dumps from the `fixtures/db` folder will be imported:

```bash
for dump in fixtures/db/*.sql; do mysql -h127.0.0.1 -P3306 -uphp-calendar-api -pphp-calendar-api --default-character-set=utf8 opa < "$dump"; done
```

or

```bash
composer db:initialize:main
```

### 5.1.5 Create dumps from database

The current database is exported to the `fixtures/db` folder:

```bash
bin/dbHelper dump
```

## 5.2 Show versions

### 5.2.1 Composer

```bash
docker compose exec -u user php php -v
```

```text
PHP 8.1.12 (cli) (built: Oct 28 2022 18:32:13) (NTS)
Copyright (c) The PHP Group
Zend Engine v4.1.12, Copyright (c) Zend Technologies
    with Zend OPcache v8.1.12, Copyright (c), by Zend Technologies
```

### 5.2.2 Composer

```bash
docker compose exec -u user php composer -V
```

```text
Composer version 2.4.4 2022-10-27 14:39:29
```

### 5.2.3 Symfony CLI

```bash
docker compose exec -u user php symfony -V
```

```text
Symfony CLI version 5.4.13 (c) 2017-2022 Symfony SAS #StandWithUkraine Support Ukraine (2022-08-16T08:17:04Z - stable)
```

### 5.2.4 Symfony Framework

```bash
docker compose exec -u user php bin/console --version
```

```text
Symfony 6.1.7 (env: dev, debug: true) #StandWithUkraine https://sf.to/ukraine
```

### 5.2.5 Application Version

```bash
docker compose exec -u user php bin/console version:show
```

```text
  Version:         0.33.0
  License:         Copyright (c) 2022 Björn Hempel
  Authors:         Björn Hempel <bjoern@hempel.li>
  PHP Version:     8.1.12
  Symfony Version: 6.1.7
```

```bash
docker compose exec -u user php bin/console version:show --format json
```

```json
{
  "version": "0.33.0",
  "license": "Copyright (c) 2022 Björn Hempel",
  "authors": [
    "Björn Hempel <bjoern@hempel.li>"
  ],
  "php-version": "8.1.12",
  "symfony-version": "6.1.7"
}
```

### 5.2.6 Node.js and npm

```bash
docker compose exec -u user php node -v
```

```text
v16.18.1
```

```bash
docker compose exec -u user php npm -v
```

```text
9.1.1
```

### 5.2.7 Yarn

```bash
docker compose exec -u user php yarn -v
```

```text
1.22.19
```

## 5.3 Docker container commands

### 5.3.1 List directories

```bash
docker compose exec -u user php ls -la
```

### 5.3.2 Show current user

```bash
docker compose exec -u user php id
```

```text
uid=1000(user) gid=1000(user) groups=1000(user)
```

## 5.4 Docker container login

### 5.4.1 PHP Container

```bash
docker compose exec -u user php bash
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## Releases

### [0.34.0] - 2022-11-23

* [#117](https://github.com/bjoern-hempel/php-calendar-api/issues/117)
  * Documentation: Version, Readme, etc.
  * Refactoring

### [0.33.0] - 2022-11-22

* [#112](https://github.com/bjoern-hempel/php-calendar-api/issues/112)
  * Add PHP rector

### [0.32.0] - 2022-11-22

* [#108](https://github.com/bjoern-hempel/php-calendar-api/issues/108)
  * Add PHP Magic Number Detector (PHPMND)

### [0.31.0] - 2022-11-21

* [#106](https://github.com/bjoern-hempel/php-calendar-api/issues/106)
  * Update to Symfony 6.1, API Platform 3.0.3 and PHP 8.1.12

### [0.30.22] - 2022-11-11

* PHPStan refactoring

### [0.30.21] - 2022-11-11

* Bump twig/twig from 3.3.8 to 3.4.3
* Disable sensiolabs security-checker

### [0.30.20] - 2022-11-11

* [#92](https://github.com/bjoern-hempel/php-calendar-api/issues/92) - Calendar: Evaluate specified location
  * Add database migration script
  * Add Croatian translation
  * Add croatian data

### [0.30.19] - 2022-07-17

* [#92](https://github.com/bjoern-hempel/php-calendar-api/issues/92) - Calendar: Evaluate specified location; Add more image properties, Connect them with specified location

### [0.30.18] - 2022-07-06

* [#92](https://github.com/bjoern-hempel/php-calendar-api/issues/92) - Calendar: Evaluate specified location; Add new design

### [0.30.17] - 2022-07-03

* [#92](https://github.com/bjoern-hempel/php-calendar-api/issues/92) - Calendar: Evaluate specified location; Do not add city and district if we do have an admin place

### [0.30.16] - 2022-07-03

* [#92](https://github.com/bjoern-hempel/php-calendar-api/issues/92) - Calendar: Evaluate specified location; Fix finding admin place (city and district) to non placeP entry

### [0.30.15] - 2022-07-03

* [#92](https://github.com/bjoern-hempel/php-calendar-api/issues/92) - Calendar: Evaluate specified location; Add admin places to search list, refactoring

### [0.30.14] - 2022-06-29

* [#92](https://github.com/bjoern-hempel/php-calendar-api/issues/92) - Calendar: Evaluate specified location; Add list index

### [0.30.13] - 2022-06-29

* [#92](https://github.com/bjoern-hempel/php-calendar-api/issues/92) - Calendar: Evaluate specified location; Add pager to result list and new relevance search

### [0.30.12] - 2022-06-28

* [#92](https://github.com/bjoern-hempel/php-calendar-api/issues/92) - Calendar: Evaluate specified location; Fix relevance search

### [0.30.11] - 2022-06-28

* [#92](https://github.com/bjoern-hempel/php-calendar-api/issues/92) - Calendar: Evaluate specified location; Add app loader

### [0.30.10] - 2022-06-26

* [#92](https://github.com/bjoern-hempel/php-calendar-api/issues/92) - Calendar: Evaluate specified location; Add tag search

### [0.30.9] - 2022-06-25

* [#92](https://github.com/bjoern-hempel/php-calendar-api/issues/92) - Calendar: Evaluate specified location; Add OpenStreetMap link

### [0.30.8] - 2022-06-25

* [#92](https://github.com/bjoern-hempel/php-calendar-api/issues/92) - Calendar: Evaluate specified location; Add current position search, Improve name search

### [0.30.7] - 2022-06-20

* [#92](https://github.com/bjoern-hempel/php-calendar-api/issues/92) - Calendar: Evaluate specified location; Add compass for places

### [0.30.6] - 2022-06-18

* [#92](https://github.com/bjoern-hempel/php-calendar-api/issues/92) - Calendar: Evaluate specified location; Add direction to place

### [0.30.5] - 2022-06-18

* [#92](https://github.com/bjoern-hempel/php-calendar-api/issues/92) - Calendar: Evaluate specified location; Remove distance from direct search (one result, id search, always 0)

### [0.30.4] - 2022-06-18

* [#92](https://github.com/bjoern-hempel/php-calendar-api/issues/92) - Calendar: Evaluate specified location; Add elevation to search result; Fix

### [0.30.3] - 2022-06-18

* [#92](https://github.com/bjoern-hempel/php-calendar-api/issues/92) - Calendar: Evaluate specified location; Add elevation to search result

### [0.30.2] - 2022-06-17

* [#92](https://github.com/bjoern-hempel/php-calendar-api/issues/92) - Calendar: Evaluate specified location; Add location parser

### [0.30.1] - 2022-06-03

* [#92](https://github.com/bjoern-hempel/php-calendar-api/issues/92) - Calendar: Evaluate specified location; Change twelvepics link

### [0.30.0] - 2022-06-03

* [#92](https://github.com/bjoern-hempel/php-calendar-api/issues/92) - Calendar: Evaluate specified location

### [0.29.0] - 2022-04-19

* [#88](https://github.com/bjoern-hempel/php-calendar-api/issues/88) - EasyAdmin: Possibility to choose the images directly in the calendar

### [0.28.0] - 2022-04-18

* [#91](https://github.com/bjoern-hempel/php-calendar-api/issues/91) - Calendar: Offer non-legal holidays

### [0.27.1] - 2022-04-18

* Bump async from 2.6.3 to 2.6.4

### [0.27.0] - 2022-04-18

* [#89](https://github.com/bjoern-hempel/php-calendar-api/issues/89) - EasyAdmin: Holiday overview within HolidayGroup

### [0.26.0] - 2022-04-09

* [#86](https://github.com/bjoern-hempel/php-calendar-api/issues/86) - Holiday Import from an API

### [0.25.1] - 2022-04-06

* Add new version of submodule submodules/bash-version-manager

### [0.25.0] - 2022-04-06

* [#94](https://github.com/bjoern-hempel/php-calendar-api/issues/94) - Docker: Add ssl to nginx

### [0.24.0] - 2022-04-04

* [#82](https://github.com/bjoern-hempel/php-calendar-api/issues/82) - EasyAdmin: Fix ajax issue on live

### [0.23.0] - 2022-04-03

* [#75](https://github.com/bjoern-hempel/php-calendar-api/issues/75) - Add yearly option to events and holidays

### [0.22.1] - 2022-04-01

* [#72](https://github.com/bjoern-hempel/php-calendar-api/issues/72) - Add root page, impress, etc.

### [0.22.0] - 2022-03-23


src/Entity/Event.php
<?php

declare(strict_types=1);

/*
 * This file is part of the bjoern-hempel/php-calendar-api project.
 *
 * (c) Björn Hempel <https://www.hempel.li/>
 *
 * For the full copyright and license information, please view the LICENSE.md
 * file that was distributed with this source code.
 */

namespace App\Entity;

use ApiPlatform\Metadata\ApiResource;
use ApiPlatform\Metadata\Delete;
use ApiPlatform\Metadata\Get;
use ApiPlatform\Metadata\GetCollection;
use ApiPlatform\Metadata\Patch;
use ApiPlatform\Metadata\Post;
use ApiPlatform\Metadata\Put;
use App\Entity\Trait\TimestampsTrait;
use App\EventListener\Entity\UserListener;
use App\Repository\EventRepository;
use App\Security\Voter\UserVoter;
use App\Utils\ArrayToObject;
use App\Utils\Traits\JsonHelper;
use DateTimeInterface;
use Doctrine\ORM\Mapping as ORM;
use Exception;
use JetBrains\PhpStorm\Pure;
use Symfony\Component\Serializer\Annotation\Groups;

/**
 * Entity class Event
 *
 * @author Björn Hempel <bjoern@hempel.li>
 * @version 0.1.3 (2022-11-11)
 * @since 0.1.3 (2022-11-21) Update to symfony 6.1
 * @since 0.1.2 (2022-11-11) PHPStan refactoring.
 * @since 0.1.1 (2022-01-29) Possibility to disable the JWT locally for debugging processes (#45)
 * @since 0.1.0 First version.
 * @package App\Entity
 */
#[ORM\Entity(repositoryClass: EventRepository::class)]
#[ORM\EntityListeners([UserListener::class])]
#[ORM\HasLifecycleCallbacks]
#[ApiResource(
    operations: [
        new GetCollection(
            normalizationContext: ['groups' => ['event']]
        ),
        new GetCollection(
            uriTemplate: '/events/extended.{_format}',
            openapiContext: [
                'description' => 'Retrieves the collection of extended Event resources.',
                'summary' => 'Retrieves the collection of extended Event resources.',
            ],
            normalizationContext: ['groups' => ['event_extended']]
        ),
        new Post(
            normalizationContext: ['groups' => ['event']],
            securityPostDenormalize: 'is_granted("'.UserVoter::ATTRIBUTE_EVENT_POST.'")',
            securityPostDenormalizeMessage: 'Only own events can be added.'
        ),

        new Delete(
            normalizationContext: ['groups' => ['event']],
            security: 'is_granted("'.UserVoter::ATTRIBUTE_EVENT_DELETE.'", object.user)',
            securityMessage: 'Only own events can be deleted.'
        ),
        new Get(
            normalizationContext: ['groups' => ['event']],
            security: 'is_granted("'.UserVoter::ATTRIBUTE_EVENT_GET.'", object.user)',
            securityMessage: 'Only own events can be read.'
        ),
        new Get(
            uriTemplate: '/events/{id}/extended.{_format}',
            openapiContext: [
                'description' => 'Retrieves an extended Event resource.',
                'summary' => 'Retrieves an extended Event resource.',
            ],
            normalizationContext: ['groups' => ['event_extended']],
            security: 'is_granted("'.UserVoter::ATTRIBUTE_EVENT_GET.'", object.user)',
            securityMessage: 'Only own events can be read.'
        ),
        new Patch(
            normalizationContext: ['groups' => ['event']],
            security: 'is_granted("'.UserVoter::ATTRIBUTE_EVENT_PATCH.'", object.user)',
            securityMessage: 'Only own events can be modified.'
        ),
        new Put(
            normalizationContext: ['groups' => ['event']],
            security: 'is_granted("'.UserVoter::ATTRIBUTE_EVENT_PUT.'", object.user)',
            securityMessage: 'Only own events can be modified.'
        )
    ],
    normalizationContext: ['enable_max_depth' => true, 'groups' => ['event']],
    order: ['id' => 'ASC'],
)]
class Event implements EntityInterface, \Stringable
{
    use TimestampsTrait;

    use JsonHelper;

    final public const CRUD_FIELDS_ADMIN = ['id', 'user'];

    final public const CRUD_FIELDS_REGISTERED = ['id', 'name', 'type', 'user', 'date', 'yearly', 'updatedAt', 'createdAt', 'configJson'];

    final public const CRUD_FIELDS_INDEX = ['id', 'name', 'type', 'user', 'date', 'yearly', 'updatedAt', 'createdAt', 'configJson'];

    final public const CRUD_FIELDS_NEW = ['id', 'name', 'type', 'user', 'date', 'yearly', 'configJson'];

    final public const CRUD_FIELDS_EDIT = self::CRUD_FIELDS_NEW;

    final public const CRUD_FIELDS_DETAIL = ['id', 'name', 'type', 'user', 'date', 'yearly', 'updatedAt', 'createdAt', 'configJson'];

    final public const CRUD_FIELDS_FILTER = ['name', 'type', 'user', 'date', 'yearly', 'updatedAt', 'createdAt'];

    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column(type: 'integer')]
    #[Groups(['event', 'event_extended'])]
    private int $id;

    #[ORM\ManyToOne(targetEntity: User::class, inversedBy: 'events')]
    #[ORM\JoinColumn(nullable: false)]
    #[Groups('event')]
    public ?User $user = null;

    #[ORM\Column(type: 'string', length: 255)]
    #[Groups(['event', 'event_extended'])]
    private string $name;

    #[ORM\Column(type: 'integer')]
    #[Groups(['event', 'event_extended'])]
    private int $type;

    #[ORM\Column(type: 'date')]
    #[Groups(['event', 'event_extended'])]
    private DateTimeInterface $date;

    /** @var array<string|int|float|bool> $config */
    #[ORM\Column(type: 'json')]
    #[Groups(['event', 'event_extended'])]
    private array $config = [
        'color' => '255,255,255,100',
    ];

    #[ORM\Column(type: 'boolean')]
    #[Groups(['event', 'event_extended'])]
    private bool $yearly = false;

    private ArrayToObject $configObject;

    /**
     * __toString method.
     *
     * @return string
     */
    public function __toString(): string
    {
        return $this->name;
    }

    /**
     * Gets the id of this event.
     *
     * @return int|null
     */
    public function getId(): ?int
    {
        return $this->id;
    }

    /**
     * Gets the user of this event.
     *
     * @return User|null
     */
    public function getUser(): ?User
    {
        return $this->user;
    }

    /**
     * Gets the user id of this event.
     *
     * @return int|null
     * @throws Exception
     */
    #[Pure]
    #[Groups(['event', 'event_extended'])]
    public function getUserId(): ?int
    {
        return $this->getUser()?->getId();
    }

    /**
     * Sets the user of this event.
     *
     * @param User|null $user
     * @return $this
     */
    public function setUser(?User $user): self
    {
        $this->user = $user;

        return $this;
    }

    /**
     * Gets the name of this event.
     *
     * @return string
     */
    public function getName(): string
    {
        return $this->name;
    }

    /**
     * Sets the name of this event.
     *
     * @param string $name
     * @return $this
     */
    public function setName(string $name): self
    {
        $this->name = $name;

        return $this;
    }

    /**
     * Gets the type of this event.
     *
     * @return int
     */
    public function getType(): int
    {
        return $this->type;
    }

    /**
     * Sets the type of this event.
     *
     * @param int $type
     * @return $this
     */
    public function setType(int $type): self
    {
        $this->type = $type;

        return $this;
    }

    /**
     * Gets the date of this event.
     *
     * @return DateTimeInterface
     */
    public function getDate(): DateTimeInterface
    {
        return $this->date;
    }

    /**
     * Sets the date of this event.
     *
     * @param DateTimeInterface $date
     * @return $this
     */
    public function setDate(DateTimeInterface $date): self
    {
        $this->date = $date;

        return $this;
    }

    /**
     * Gets the config.
     *
     * @return array<string|int|float|bool>
     */
    public function getConfig(): array
    {
        return $this->config;
    }

    /**
     * Gets the config as object.
     *
     * @return ArrayToObject
     * @throws Exception
     */
    public function getConfigObject(): ArrayToObject
    {
        if (!isset($this->configObject)) {
            $this->configObject = new ArrayToObject($this->config);
        }

        return $this->configObject;
    }

    /**
     * Sets the config.
     *
     * @param array<string|int|float|bool> $config
     * @return $this
     * @throws Exception
     */
    public function setConfig(array $config): self
    {
        $this->config = $config;

        $this->configObject = new ArrayToObject($config);

        return $this;
    }

    /**
     * Gets the config element as JSON.
     *
     * @param bool $beautify
     * @return string
     * @throws Exception
     */
    public function getConfigJson(bool $beautify = true): string
    {
        return self::jsonEncode($this->config, $beautify, 2);
    }

    /**
     * Sets the config element from JSON.
     *
     * @param string $json
     * @return $this
     */
    public function setConfigJson(string $json): self
    {
        $this->config = self::jsonDecodeArray($json);

        return $this;
    }

    /**
     * Gets the config element as JSON.
     *
     * @param bool $beautify
     * @return string
     * @throws Exception
     */
    public function getConfigJsonRaw(bool $beautify = true): string
    {
        return $this->getConfigJson(false);
    }

    /**
     * Sets the config element from JSON.
     *
     * @param string $json
     * @return $this
     */
    public function setConfigJsonRaw(string $json): self
    {
        return $this->setConfigJson($json);
    }

    /**
     * Gets the yearly status of this holiday.
     *
     * @return bool|null
     */
    public function getYearly(): ?bool
    {
        return $this->yearly;
    }

    /**
     * Sets the yearly status from this holiday.
     *
     * @param bool $yearly
     * @return $this
     */
    public function setYearly(bool $yearly): self
    {
        $this->yearly = $yearly;

        return $this;
    }
}
